src/coreclr/tools/ILVerification/ILImporter.Verify.cs

```diff
@@ -1119,37 +1119,108 @@ void CheckDelegateCreation(StackValue ftn, StackValue obj)
+
             // Compare signature parameters
-            if (ftnSignature.Length != delegateSignature.Length)
+
+            bool isOpenDelegate;
+            if (totalTargetArgs == invokeSignature.Length)
+            {
+                // All arguments provided by invoke, delegate must be open.
+                isOpenDelegate = true;
+            }
+            else if (totalTargetArgs == invokeSignature.Length + 1)
+            {
+                // One too few arguments provided by invoke, delegate must be closed.
+                isOpenDelegate = false;
+            }
+            else
+            {
                 return false;
+            }
 
-            for (int i = 0; i < ftnSignature.Length; i++)
+            // An open static delegate which takes no arguments. In that case we're done.
+            if (totalTargetArgs == 0)
             {
-                if (!IsAssignable(delegateSignature[i], ftnSignature[i]))
+                Debug.Assert(isOpenDelegate);
+                return true;
+            }
+
+            int consumedArgs = 0;
+
+            TypeDesc firstInvokeArg;
+            if (isOpenDelegate)
+            {
+                // If we're looking at an open delegate but the caller has provided a target it's not a match.
+                if (firstArg != null)
                     return false;
+
+                firstInvokeArg = invokeSignature[0];
+                consumedArgs++;
             }
+            else
+            {
+                // If we're looking at a closed delegate but the caller has not provided a target it's not a match.
+                if (firstArg == null)
+                    return false;
 
-            // Compare return type
-            return IsAssignable(ftnSignature.ReturnType, delegateSignature.ReturnType);
+                firstInvokeArg = firstArg;
+            }
+
+            TypeDesc firstTargetArg;
+            if (targetSignature.IsStatic)
+            {
+                // Checked above
+                Debug.Assert(targetSignature.Length != 0);
+
+                // The first argument for a static method is the first fixed arg.
+                firstTargetArg = targetSignature[0];
+                consumedArgs--;
+            }
+            else
+            {
+                // The type of the first argument to an instance method is from the method type.
+                firstTargetArg = targetMethod.OwningType;
+
+                // If the delegate is open and the target method is on a value type or primitive then the first argument of the invoke
+                // method must be a reference to that type. So make promote the type we got from the reference to a ref. (We don't need to
+                // do this for the closed instance case because there we got the invocation side type from the first arg passed in, i.e.
+                // it's had the ref stripped from it implicitly).
```

Comment:
nit: looks like this `comdelegate.cpp` comment can be simplified a little
```suggestion
                // method must be a reference to that type. So make the type we got from the reference to a ref. (We don't need to
                // do this for the closed instance case because there we got the invocation side type from the first arg passed in, i.e.
                // ref was stripped from it implicitly).
```